## RC Playlists

In [1]:
import pandas as pd
import numpy as np
import json
from json.decoder import JSONDecodeError
from time import sleep

import spotipy
import spotipy.util as util
import spotipy.oauth2 as oauth2
#from spotipy.oauth2 import SpotifyClientCredentials # might not need this because above imports all

In [78]:
client_id='YOUR_CLIENT_ID'
client_secret='YOUR_CLIENT_SECRET'
redirect_uri='http://your_redirect.com/'

# To authorize your API calls without going through the headache 
# of verifying the redirect URL. Just make use of the oauth2 method:
# https://github.com/plamere/spotipy/issues/194

credentials = oauth2.SpotifyClientCredentials(
        client_id=client_id,
        client_secret=client_secret)

token = credentials.get_access_token()
sp = spotipy.Spotify(auth=token)

sp.trace = False # turns on tracing - this gives a whole bunch of SHIT that you don't need!
sp.trace_out = False # turns on trace out

In [3]:
user = 'spotify:user:USERID'

pl_1 = 'spotify:user:USER:playlist:ID1'
pl_2 = 'spotify:user:USER:playlist:ID1'

# returns 100 tracks at a time
pl_1_data = sp.user_playlist_tracks(user, pl_1)

In [50]:
### ALBUM INFO
# album uri
pl_1_data['items'][0]['track']['album']['uri']

# album release date
pl_1_data['items'][0]['track']['album']['release_date']

### ARTIST INFO
# artist name
pl_1_data['items'][0]['track']['artists'][0]['name']

# uri
pl_1_data['items'][0]['track']['artists'][0]['uri']


### TRACK INFO
# track name
pl_1_data['items'][0]['track']['name']

# track uri
pl_1_data['items'][0]['track']['uri']

# track popularity
pl_1_data['items'][0]['track']['popularity']



7

# Get ALL tracks from playlist

In [ ]:
offset = 0

data = [1,1,1]

new = []

while offset < 10:
    for i in data:
        
        new.append(data[i] + 1)
        offset += 1



# loop through playlist to extract song data
user = 'spotify:user:USERID'

pl_1 = 'spotify:user:USER:playlist:ID1'
pl_2 = 'spotify:user:USER:playlist:ID1'

# returns 100 tracks at a time
#pl_1_data = sp.user_playlist_tracks(user, pl_1, offset=100)['items']

data = []
data_2 = []
offset = 0


while offset < 10000:

    playlist = sp.user_playlist_tracks(user, pl_2, offset=offset)['items']

    for item in playlist:

        album_uri = item['track']['album']['uri']
        release_date = item['track']['album']['release_date']
        artist_name = item['track']['artists'][0]['name']
        artist_uri = item['track']['artists'][0]['uri']

        track_name = item['track']['name']
        track_uri = item['track']['uri']
        popularity = item['track']['popularity']

        data_2.append({
            'album_uri': album_uri,
            'release_date': release_date,
            'artist_name': artist_name,
            'artist_uri': artist_uri,
            'track_name': track_name,
            'track_uri': track_uri,
            'popularity':popularity
        })

    offset += 100
    print(len(data_2))

In [ ]:
df_1 = pd.DataFrame(data, columns=['album_uri', 'release_date', 'artist_name', 'artist_uri',
                                  'track_name', 'track_uri', 'popularity'])

df_1.drop_duplicates(['track_uri'], inplace=True)

df_1.to_csv('df_1.csv', index=False)

df_1.info()

In [150]:
df_2 = pd.DataFrame(data_2, columns=['album_uri', 'release_date', 'artist_name', 'artist_uri',
                                  'track_name', 'track_uri', 'popularity'])

df_2.drop_duplicates(['track_uri'], inplace=True)

df_2.to_csv('df_2.csv', index=False)

df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9302 entries, 0 to 9303
Data columns (total 7 columns):
album_uri       9302 non-null object
release_date    9302 non-null object
artist_name     9302 non-null object
artist_uri      9302 non-null object
track_name      9302 non-null object
track_uri       9302 non-null object
popularity      9302 non-null int64
dtypes: int64(1), object(6)
memory usage: 581.4+ KB


In [156]:
df_master = pd.concat([df_1, df_2], ignore_index=True)

df_master.to_csv('df_master.csv', index=False)

df_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19298 entries, 0 to 19297
Data columns (total 7 columns):
album_uri       19298 non-null object
release_date    19298 non-null object
artist_name     19298 non-null object
artist_uri      19298 non-null object
track_name      19298 non-null object
track_uri       19298 non-null object
popularity      19298 non-null int64
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


# Get Audio Features

In [216]:
df_master['track_id'] = df_master.track_uri.str.extract('(?<=spotify:track:)(.*)', expand=False)

df_master.to_csv('df_master.csv', index=False)

df_master.head()

album_uri release_date   artist_name  \
0  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   
1  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   
2  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   
3  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   
4  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   

                              artist_uri     track_name  \
0  spotify:artist:1ruRQnflSHJVLqvqRe5MRB       Dog Days   
1  spotify:artist:1ruRQnflSHJVLqvqRe5MRB            Bug   
2  spotify:artist:1ruRQnflSHJVLqvqRe5MRB      Clara Bow   
3  spotify:artist:1ruRQnflSHJVLqvqRe5MRB  Long Painting   
4  spotify:artist:1ruRQnflSHJVLqvqRe5MRB     Glory Weed   

                              track_uri  popularity                track_id  
0  spotify:track:799DkBhHRSdf4AhRnvs93t           4  799DkBhHRSdf4AhRnvs93t  
1  spotify:track:0gqGR7gQXhmL0EtLHxKyu6           6  0gqGR7gQXhmL0EtLHxKyu6  
2  spotify:track:42GP7EH07xO0cYdRGft1NM           8  42GP7EH07xO0cYdRGft1NM  
3  spotify:track:4nsxTK5WpbwLRkuRJmR6mJ           5  4nsxTK5WpbwLRkuRJmR6mJ  
4  spotify:track:4KNNyWjbbH7ikafrhQm7dK           3  4KNNyWjbbH7ikafrhQm7dK

In [80]:
df_master.query('track_uri == "spotify:track:6C14iBEiZPeKhACrUGMfE7"')

#df_master[1427:1429]

album_uri release_date artist_name  \
6445  spotify:album:7oe03yjIB4642gyQSTI1rF         1992     Rubicon   

                                 artist_uri         track_name  \
6445  spotify:artist:4k3iS4bsVsTi9dIq2ybtWK  What Starts, Ends   

                                 track_uri  popularity                track_id  
6445  spotify:track:6C14iBEiZPeKhACrUGMfE7           1  6C14iBEiZPeKhACrUGMfE7

### Split Track ID list into sparts

This will make it easier to loop through and not have to baby sit the extraction as much.


In [154]:
# create a list of track ids from the sp_track_df
track_id_list = df_master.track_uri.tolist()

l = track_id_list[6446:]

def split(l,n):
    return [l[i::n] for i in range(n)]

track_id_split = split(l,10)

len(track_id_split)

10

In [155]:
# check the split worked

print(len(track_id_list[6446:]))
print()
print([len(i) for i in track_id_split])
print()
print(sum(len(i) for i in track_id_split))

12852

[1286, 1286, 1285, 1285, 1285, 1285, 1285, 1285, 1285, 1285]

12852


### Extract data from API

- Loop through list of id lists
- set credentials for each new list
- extract features
- append to empty list 
- save as df

- each df should include the last df plus all new records
- add print statements to keep track of progress

In [167]:
### TEST
len(track_id_split)

10

In [170]:
len(features_data)
#len(features_df.id)

10495

In [171]:
# set client id, secret and redirect uri
client_id='YOUR_CLIENT_ID'
client_secret='YOUR_CLIENT_SECRET'
redirect_uri='http://your_redirect.com/'

# column names for DF
cols = ['id','uri','analysis_url','track_href','type','duration_ms',
        'acousticness','danceability','energy','instrumentalness','liveness',
        'loudness','speechiness','valence','key','mode','tempo','time_signature']

# new column names for DF
new_cols = ['id', 'track_uri', 'analysis_url', 'track_href', 'type', 'duration_ms',
       'acousticness', 'danceability', 'energy', 'instrumentalness',
       'liveness', 'loudness', 'speechiness', 'valence', 'key', 'mode',
       'tempo', 'time_signature']


# empty list to append to
features_data = []

# counter variable
# increase with each loop
count = 9


for id_list in track_id_split[8:10]:
    
    import spotipy
    import spotipy.util as util
    import spotipy.oauth2 as oauth2
    
    ### SET CREDENTIALS ###
    credentials = oauth2.SpotifyClientCredentials(client_id=client_id, 
                                                  client_secret=client_secret)
    
    token = credentials.get_access_token()
    sp = spotipy.Spotify(auth=token)
    
    sp.trace = False
    sp.trace_out = False
    
    print('Credentials set')
    print('This is loop number {}'.format(count))
    
    ### extract features
    for track in id_list:
          
        try:
            # extract features
            features = sp.audio_features(track)

            # append to features_data
            features_data.append(features[0])

            # print(features_data[-1]['uri'], features_data[-1]['energy'], len(features_data))

        except (TypeError, AttributeError):
            print('error: ',track)

        sleep(1) # don't be a jerk
    
    ### CREATE DF ###
    print('creating DF')
    features_df = pd.DataFrame(features_data, columns=cols)
    
    # rename columns
    features_df.columns = new_cols
    
    # save the df
    f_name = 'rc_features_df_{}.csv'.format(count)
    
    features_df.to_csv(f_name, index=False)
    
    print(len(features_df['id']))
    print('new df created')
    
    # add 1 too the loop counter
    count += 1


Credentials set
This is loop number 9
creating DF
1285
new df created
Credentials set
This is loop number 10
creating DF
2570
new df created


In [159]:
features_df.track_uri.iloc[-1]

'spotify:track:1VA4wbVvDbarfzzdoHxSFb'

### Create Features DF

In [83]:
cols = ['id','uri','analysis_url','track_href','type','duration_ms',
        'acousticness','danceability','energy','instrumentalness','liveness',
        'loudness','speechiness','valence','key','mode','tempo','time_signature']

features_df = pd.DataFrame(features_data, columns=cols)

# rename columns
new_cols = ['id', 'track_uri', 'analysis_url', 'track_href', 'type', 'duration_ms',
       'acousticness', 'danceability', 'energy', 'instrumentalness',
       'liveness', 'loudness', 'speechiness', 'valence', 'key', 'mode',
       'tempo', 'time_signature']

features_df.columns = new_cols

# save DF
# save df to disk
features_df.to_csv('rc_features_df_05.csv', index=False)

# save df to disk
features_df.to_csv('rc_features_df_06.csv', index=False)

features_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2238 entries, 0 to 2237
Data columns (total 18 columns):
id                  2238 non-null object
track_uri           2238 non-null object
analysis_url        2238 non-null object
track_href          2238 non-null object
type                2238 non-null object
duration_ms         2238 non-null int64
acousticness        2238 non-null float64
danceability        2238 non-null float64
energy              2238 non-null float64
instrumentalness    2238 non-null float64
liveness            2238 non-null float64
loudness            2238 non-null float64
speechiness         2238 non-null float64
valence             2238 non-null float64
key                 2238 non-null int64
mode                2238 non-null int64
tempo               2238 non-null float64
time_signature      2238 non-null int64
dtypes: float64(9), int64(4), object(5)
memory usage: 314.8+ KB


## Merge Data

### Mereg Features

In [2]:
df_master = pd.read_csv('df_features_master.csv')
df_f_01 = pd.read_csv('rc_features_df_1.csv')
df_f_02 = pd.read_csv('rc_features_df_2.csv')
df_f_03 = pd.read_csv('rc_features_df_3.csv')
df_f_04 = pd.read_csv('rc_features_df_4.csv')
df_f_05 = pd.read_csv('rc_features_df_5.csv')
df_f_06 = pd.read_csv('rc_features_df_6.csv')
df_f_07 = pd.read_csv('rc_features_df_7.csv')
df_f_08 = pd.read_csv('rc_features_df_8.csv')
df_f_09 = pd.read_csv('rc_features_df_9.csv')
df_f_10 = pd.read_csv('rc_features_df_10.csv')


df_features_master = pd.concat([df_master, df_f_01, df_f_02, df_f_03, df_f_04, df_f_05, df_f_06, df_f_07, df_f_08, df_f_09, df_f_10])

# drop duplicates
df_features_master.drop_duplicates(keep='first')

df_features_master.to_csv('df_features_master_02.csv', index=False)

df_features_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56557 entries, 0 to 2569
Data columns (total 18 columns):
id                  56557 non-null object
track_uri           56557 non-null object
analysis_url        56557 non-null object
track_href          56557 non-null object
type                56557 non-null object
duration_ms         56557 non-null int64
acousticness        56557 non-null float64
danceability        56557 non-null float64
energy              56557 non-null float64
instrumentalness    56557 non-null float64
liveness            56557 non-null float64
loudness            56557 non-null float64
speechiness         56557 non-null float64
valence             56557 non-null float64
key                 56557 non-null int64
mode                56557 non-null int64
tempo               56557 non-null float64
time_signature      56557 non-null int64
dtypes: float64(9), int64(4), object(5)
memory usage: 8.2+ MB


In [182]:
df_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19298 entries, 0 to 19297
Data columns (total 8 columns):
album_uri       19298 non-null object
release_date    19298 non-null object
artist_name     19298 non-null object
artist_uri      19298 non-null object
track_name      19298 non-null object
track_uri       19298 non-null object
popularity      19298 non-null int64
track_id        19298 non-null object
dtypes: int64(1), object(7)
memory usage: 1.2+ MB


In [188]:
# import master playlist data
#df_master = pd.read_csv('df_master.csv')


df_master_merged = pd.merge(left=df_master, right=df_features_master, how='left', on='track_uri')

#df_master_merged.to_csv('df_maser_2.csv', index=False)

df_master_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59368 entries, 0 to 59367
Data columns (total 25 columns):
album_uri           59368 non-null object
release_date        59368 non-null object
artist_name         59368 non-null object
artist_uri          59368 non-null object
track_name          59368 non-null object
track_uri           59368 non-null object
popularity          59368 non-null int64
track_id            59368 non-null object
id                  57142 non-null object
analysis_url        57142 non-null object
track_href          57142 non-null object
type                57142 non-null object
duration_ms         57142 non-null float64
acousticness        57142 non-null float64
danceability        57142 non-null float64
energy              57142 non-null float64
instrumentalness    57142 non-null float64
liveness            57142 non-null float64
loudness            57142 non-null float64
speechiness         57142 non-null float64
valence             57142 non-null float64
k

In [186]:
df_master_merged.tail()

album_uri release_date  artist_name  \
59363  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
59364  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
59365  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
59366  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
59367  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   

                                  artist_uri    track_name  \
59363  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi  Here You Are   
59364  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi  Here You Are   
59365  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi  Here You Are   
59366  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi  Here You Are   
59367  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi  Here You Are   

                                  track_uri  popularity  \
59363  spotify:track:1DTc8zGuYwi16ewezwoQK8          35   
59364  spotify:track:1DTc8zGuYwi16ewezwoQK8          35   
59365  spotify:track:1DTc8zGuYwi16ewezwoQK8          35   
59366  spotify:track:1DTc8zGuYwi16ewezwoQK8          35   
59367  spotify:track:1DTc8zGuYwi16ewezwoQK8          35   

                     track_id                      id  \
59363  1DTc8zGuYwi16ewezwoQK8  1DTc8zGuYwi16ewezwoQK8   
59364  1DTc8zGuYwi16ewezwoQK8  1DTc8zGuYwi16ewezwoQK8   
59365  1DTc8zGuYwi16ewezwoQK8  1DTc8zGuYwi16ewezwoQK8   
59366  1DTc8zGuYwi16ewezwoQK8  1DTc8zGuYwi16ewezwoQK8   
59367  1DTc8zGuYwi16ewezwoQK8  1DTc8zGuYwi16ewezwoQK8   

                                            analysis_url       ...        \
59363  https://api.spotify.com/v1/audio-analysis/1DTc...       ...         
59364  https://api.spotify.com/v1/audio-analysis/1DTc...       ...         
59365  https://api.spotify.com/v1/audio-analysis/1DTc...       ...         
59366  https://api.spotify.com/v1/audio-analysis/1DTc...       ...         
59367  https://api.spotify.com/v1/audio-analysis/1DTc...       ...         

      energy instrumentalness  liveness  loudness  speechiness  valence  key  \
59363  0.423            0.939    0.0889   -11.934       0.0496    0.272  4.0   
59364  0.423            0.939    0.0889   -11.934       0.0496    0.272  4.0   
59365  0.423            0.939    0.0889   -11.934       0.0496    0.272  4.0   
59366  0.423            0.939    0.0889   -11.934       0.0496    0.272  4.0   
59367  0.423            0.939    0.0889   -11.934       0.0496    0.272  4.0   

       mode    tempo  time_signature  
59363   1.0  117.551             4.0  
59364   1.0  117.551             4.0  
59365   1.0  117.551             4.0  
59366   1.0  117.551             4.0  
59367   1.0  117.551             4.0  

[5 rows x 25 columns]

# MERGE DATA

In [63]:
# DF 1
# DF 2
# DF Master

#df_1 = pd.read_csv('df_1.csv')
#df_2 = pd.read_csv('df_2.csv')

df_master = pd.read_csv('df_master.csv')

In [11]:
a = len(df_1.album_uri)

b = len(df_2.album_uri)

print((a + b), len(df_master.album_uri))

19298 19298


In [64]:
# FEATURES 2

df_features_master_2 = pd.read_csv('df_features_master_02.csv')

In [28]:
df_features_master_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56557 entries, 0 to 56556
Data columns (total 18 columns):
id                  56557 non-null object
track_uri           56557 non-null object
analysis_url        56557 non-null object
track_href          56557 non-null object
type                56557 non-null object
duration_ms         56557 non-null int64
acousticness        56557 non-null float64
danceability        56557 non-null float64
energy              56557 non-null float64
instrumentalness    56557 non-null float64
liveness            56557 non-null float64
loudness            56557 non-null float64
speechiness         56557 non-null float64
valence             56557 non-null float64
key                 56557 non-null int64
mode                56557 non-null int64
tempo               56557 non-null float64
time_signature      56557 non-null int64
dtypes: float64(9), int64(4), object(5)
memory usage: 7.8+ MB


### drop any duplicates

- features df is way too long, must be some duplicates

In [65]:
df_features_master_2['id'].duplicated().any()

df_features_master_2.drop_duplicates(keep='first', inplace=True)

df_features_master_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18496 entries, 0 to 56556
Data columns (total 18 columns):
id                  18496 non-null object
track_uri           18496 non-null object
analysis_url        18496 non-null object
track_href          18496 non-null object
type                18496 non-null object
duration_ms         18496 non-null int64
acousticness        18496 non-null float64
danceability        18496 non-null float64
energy              18496 non-null float64
instrumentalness    18496 non-null float64
liveness            18496 non-null float64
loudness            18496 non-null float64
speechiness         18496 non-null float64
valence             18496 non-null float64
key                 18496 non-null int64
mode                18496 non-null int64
tempo               18496 non-null float64
time_signature      18496 non-null int64
dtypes: float64(9), int64(4), object(5)
memory usage: 2.7+ MB


## MERGE features and master DF

- Master_df is 19298 long
- Features_df is 18496 long

- should end up being 1928 long with some NA

In [43]:
df_master.head(2)

album_uri release_date   artist_name  \
0  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   
1  spotify:album:1MAvStlhPSIZ8ubHyeek10   2004-01-01  50 Foot Wave   

                              artist_uri track_name  \
0  spotify:artist:1ruRQnflSHJVLqvqRe5MRB   Dog Days   
1  spotify:artist:1ruRQnflSHJVLqvqRe5MRB        Bug   

                              track_uri  popularity                track_id  
0  spotify:track:799DkBhHRSdf4AhRnvs93t           4  799DkBhHRSdf4AhRnvs93t  
1  spotify:track:0gqGR7gQXhmL0EtLHxKyu6           6  0gqGR7gQXhmL0EtLHxKyu6

In [44]:
df_features_master_2.head(2)

id                             track_uri  \
0  799DkBhHRSdf4AhRnvs93t  spotify:track:799DkBhHRSdf4AhRnvs93t   
1  0gqGR7gQXhmL0EtLHxKyu6  spotify:track:0gqGR7gQXhmL0EtLHxKyu6   

                                        analysis_url  \
0  https://api.spotify.com/v1/audio-analysis/799D...   
1  https://api.spotify.com/v1/audio-analysis/0gqG...   

                                          track_href            type  \
0  https://api.spotify.com/v1/tracks/799DkBhHRSdf...  audio_features   
1  https://api.spotify.com/v1/tracks/0gqGR7gQXhmL...  audio_features   

   duration_ms  acousticness  danceability  energy  instrumentalness  \
0       231600      0.000056         0.244   0.943           0.00541   
1       260027      0.000531         0.263   0.967           0.00023   

   liveness  loudness  speechiness  valence  key  mode    tempo  \
0     0.222    -4.542       0.1050    0.420    9     0  108.877   
1     0.441    -3.334       0.0896    0.298    0     1  121.016   

   time_signature  
0               3  
1               4

In [66]:
db_master = pd.merge(df_master, df_features_master_2, on='track_uri', how='left')

db_master.drop_duplicates(keep='first', inplace=True)

db_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20722 entries, 0 to 20872
Data columns (total 25 columns):
album_uri           20722 non-null object
release_date        20722 non-null object
artist_name         20722 non-null object
artist_uri          20722 non-null object
track_name          20722 non-null object
track_uri           20722 non-null object
popularity          20722 non-null int64
track_id            20722 non-null object
id                  18496 non-null object
analysis_url        18496 non-null object
track_href          18496 non-null object
type                18496 non-null object
duration_ms         18496 non-null float64
acousticness        18496 non-null float64
danceability        18496 non-null float64
energy              18496 non-null float64
instrumentalness    18496 non-null float64
liveness            18496 non-null float64
loudness            18496 non-null float64
speechiness         18496 non-null float64
valence             18496 non-null float64
k

In [68]:
db_master.to_csv('df_master_meged_py.csv', index=False)

db_master.tail()

album_uri release_date  artist_name  \
20868  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
20869  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
20870  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
20871  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   
20872  spotify:album:1PP5mb8VOZfofwDzHRQPYh   2018-03-16  Yo La Tengo   

                                  artist_uri              track_name  \
20868  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi  What Chance Have I Got   
20869  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi         Esportes Casual   
20870  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi                 Forever   
20871  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi         Out of the Pool   
20872  spotify:artist:5hAhrnb0Ch4ODwWu4tsbpi            Here You Are   

                                  track_uri  popularity  \
20868  spotify:track:1VA4wbVvDbarfzzdoHxSFb          36   
20869  spotify:track:4JqxXC971HEhbH5Idn9N7k          35   
20870  spotify:track:5FVXLaVVmAhdfHa7HZbhC3          33   
20871  spotify:track:2FNfZxFfvwYIwkqM2cUECd          35   
20872  spotify:track:1DTc8zGuYwi16ewezwoQK8          35   

                     track_id                      id  \
20868  1VA4wbVvDbarfzzdoHxSFb  1VA4wbVvDbarfzzdoHxSFb   
20869  4JqxXC971HEhbH5Idn9N7k  4JqxXC971HEhbH5Idn9N7k   
20870  5FVXLaVVmAhdfHa7HZbhC3  5FVXLaVVmAhdfHa7HZbhC3   
20871  2FNfZxFfvwYIwkqM2cUECd  2FNfZxFfvwYIwkqM2cUECd   
20872  1DTc8zGuYwi16ewezwoQK8  1DTc8zGuYwi16ewezwoQK8   

                                            analysis_url       ...        \
20868  https://api.spotify.com/v1/audio-analysis/1VA4...       ...         
20869  https://api.spotify.com/v1/audio-analysis/4Jqx...       ...         
20870  https://api.spotify.com/v1/audio-analysis/5FVX...       ...         
20871  https://api.spotify.com/v1/audio-analysis/2FNf...       ...         
20872  https://api.spotify.com/v1/audio-analysis/1DTc...       ...         

      energy instrumentalness  liveness  loudness  speechiness  valence  key  \
20868  0.357            0.482    0.1020   -15.933       0.0523    0.331  2.0   
20869  0.524            0.880    0.1210   -11.169       0.0380    0.802  6.0   
20870  0.336            0.561    0.4050   -18.830       0.0428    0.420  2.0   
20871  0.668            0.874    0.1390   -10.754       0.0341    0.843  1.0   
20872  0.423            0.939    0.0889   -11.934       0.0496    0.272  4.0   

       mode    tempo  time_signature  
20868   1.0  184.505             4.0  
20869   0.0   91.154             4.0  
20870   1.0   78.456             4.0  
20871   0.0   98.041             4.0  
20872   1.0  117.551             4.0  

[5 rows x 25 columns]

# CHECK AGAINST EXCEL MERGE

In [61]:
df_merged = pd.read_csv('df_master_meged.csv')
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19298 entries, 0 to 19297
Data columns (total 24 columns):
album_uri           19298 non-null object
release_date        19298 non-null object
artist_name         19298 non-null object
artist_uri          19298 non-null object
track_name          19298 non-null object
track_uri           19298 non-null object
popularity          19298 non-null int64
track_id            19298 non-null object
analysis_url        17072 non-null object
track_href          17072 non-null object
type                17072 non-null object
duration_ms         17072 non-null float64
acousticness        17072 non-null float64
danceability        17072 non-null float64
energy              17072 non-null float64
instrumentalness    17072 non-null float64
liveness            17072 non-null float64
loudness            17072 non-null float64
speechiness         17072 non-null float64
valence             17072 non-null float64
key                 17072 non-null float64


## CHECK ARE EQUAL

In [69]:
db_master.equals(df_merged)

False